stats.lua: fixes and unification

christoph-heinrich opened the pull request:
Fixes scrolling for page 0 in the terminal and unifies the scrolling implementation.
Also introduces a new line shortening function for the terminal that is used by all scrollable pages now.

kasper93 commented:
Could we detect if it fits on screen and make it scrollable only then? Sounds like hard thing to do, with all font sizes and dpi.

I don't like enabling it when not necessary, because then we steal arrow input for no reason. I wanted to mention it, but not big deal probably.

christoph-heinrich commented:
> Could we detect if it fits on screen and make it scrollable only then? Sounds like hard thing to do, with all font sizes and dpi.

Yes that should be possible, that's what we're doing in the console.

christoph-heinrich commented:
After thinking about it for a bit, we could use [compute_bounds](https://mpv.io/manual/master/#command-interface-compute-bounds) to know if the text goes beyond the screen and then let it scroll until it doesn't clip anymore.

avih commented:
> Could we detect if it fits on screen and make it scrollable only then?

Not sure about that, even if possible, I'm not thrilled by a page "randomly" hijacking or not hijacking the arrow keys.

Keep in mind that the scroll keys are configurable, so if hijacking arrows become an issue, a user could always map it to something else.

I'm in favor of keeping it simple. It's already not very simple, so IMO it would be better to not automate it further.

christoph-heinrich commented:
I have yet to look into limiting ass scrolling via compute_bounds, but other then that it should be pretty good now.
The default page shows 22 lines and the other scroll-able pages show 23 lines. That is because the default page doesn't
have a fixed header, and so the new line at the beginning of each but the front line adds an additional line in the output.
The solution to that would be to not bake in new lines and instead add them during concatenation. I haven't don't done that yet because I suspect it requires a lot of changes and the way it is right now is already good enough.

avih commented:
Regarding the terminal truncation, you're obviously aware of the following issues:
- Tabs are counted incorrectly unless they're at 0 or tabstop.
- SGR sequences are assumed always 4 bytes.
- Every UTF8 byte counts as a terminal cell.
- Truncation can happen inside an SGR sequence.
- If truncation happens after an SGR sequence but before the matching SGR reset, then the SGR will still be in effect after the truncation. Terminals are especially bad at this if the first SGR changed the BG color. Do keep in mind that the SGR sequences are user-configurable.

Additionally, surely some users might try to resize the terminal to avoid the truncation... to no avail. So preferably it should either find the terminal width automatically, or take it from some configuration value.

Here's an alternative solution which addresses all those issues (sans finding the term width):
```lua
-- assumptions:
--   s is composed of SGR escape sequences and/or printable UTF8 sequences
--   printable codepoints occupy one terminal cell (we don't have wcwidth)
--   tabstops are 8, 16, 24..., and the output starts at 0 or a tabstop
function term_width(s)
    local w = 0
    for c in s:gsub("\27.-m", ""):gsub("\27.*", ""):gmatch(".") do
        -- tab counts to the next stop, UTF8 continuations ignored, else 1
        local x = c:byte(1)
        w = w + (x == 9 and 8 - w % 8 or x > 127 and x < 192 and 0 or 1)
    end
    return w
end

-- if maxwidth <= 2 and s doesn't fit then the output is "..." (more than 2)
function term_ellipsis(s, maxwidth)
    if term_width(s) > maxwidth then
        repeat s = s:sub(1, #s - 1)
        until  #s == 0 or term_width(s) <= maxwidth - 3
        s = s .. (s:find("\27") and "\27[m..." or "...")
    end
    return s
end
```

If you think this too slow because it's O(N^2), then after you confirm that with some measurements, you can use this function instead, which does that in O(N log N), and is probably good enough for most practical use cases:
```lua
-- if maxwidth <= 2 and s doesn't fit then the output is "..." (more than 2)
function term_ellipsis(s, maxwidth)
    if term_width(s) > maxwidth then
        -- binary search for the longest prefix which fits maxwidth
        local low, high = 0, #s  -- low = 0 in case s is "\t..."
        while low < high-1 do
            local mid = (low + high) / 2
            if term_width(s:sub(1, mid)) <= maxwidth - 3 then
                low = mid
            else
                high = mid
            end
        end
        s = s:sub(1, low)
        s = s .. (s:find("\27") and "\27[m..." or "...")
    end
    return s
end
```

If you still think this is too slow, then feel free to provide a good solution in O(N), after you measure how much you gain from it.

christoph-heinrich commented:
Benchmarks were done with
<details>
<summary>benchmarking diff</summary>

```diff
diff --git a/player/lua/stats.lua b/player/lua/stats.lua
index 7873259fbf..f3b8f2c213 100644
--- a/player/lua/stats.lua
+++ b/player/lua/stats.lua
@@ -843,24 +843,73 @@ local function eval_ass_formatting()
     end
 end
 
-local function truncate_lines(content, from, to)
-    local tab = string.byte("\t")
-    for i = from, to do
-        local line = content[i]
-        local tabs = 0
-        for j = 1, #line do
-            if line:byte(j) == tab then
-                tabs = tabs + 1
-            end
-        end
-        local pos, esc = 1, -1
-        repeat
-            _, pos = line:find("\027[", pos, true)
-            esc = esc + 1
-        until pos == nil or pos + tabs * 7 - esc * 4 >= 80
-        content[i] = ellipsis(line, 80 - tabs * 7 + esc * 4)
-    end
-end
+-- function term_ellipsis(s, maxwidth)
+--     -- \t is 9, m is 109, [ is 91 and UTF-8 continuations are [128,191]
+--     local w, i, l, i0, i1, i2, i3 = 0, 1, s:len(), 0, 0, 0, 0
+--     while w <= maxwidth and i <= l do
+--         local c, to = s:byte(i), nil
+--         if c == 27 and s:byte(i + 1) == 91 then
+--             to = s:byte(i + 2) == 109 and i + 2 or s:byte(i + 3) == 109 and i + 3
+--                 or s:byte(i + 4) == 109 and i + 4 or s:byte(i + 5) == 109 and i + 5
+--             if to then
+--                 i = to
+--             end
+--         end
+--         local cs = (c == 9 and 8 - w % 8 or to and 0 or c > 127 and c < 192 and 0 or 1)
+--         w = w + cs
+--         if cs > 0 then
+--             i3, i2, i1, i0 = i2, i1, i0, i
+--         end
+--         i = i + 1
+--     end
+--     if w > maxwidth then
+--         s = s:sub(1, i3 > 1 and i3 - 1 or 0) .. "\27[0m..."
+--     end
+--     return s
+-- end
+
+-- assumptions:
+--   s is composed of SGR escape sequences and/or printable UTF8 sequences
+--   printable codepoints occupy one terminal cell (we don't have wcwidth)
+--   tabstops are 8, 16, 24..., and the output starts at 0 or a tabstop
+-- function term_width(s)
+--     local w = 0
+--     for c in s:gsub("\27.-m", ""):gsub("\27.*", ""):gmatch(".") do
+--         -- tab counts to the next stop, UTF8 continuations ignored, else 1
+--         local x = c:byte(1)
+--         w = w + (x == 9 and 8 - w % 8 or x > 127 and x < 192 and 0 or 1)
+--     end
+--     return w
+-- end
+
+-- if maxwidth <= 2 and s doesn't fit then the output is "..." (more than 2)
+-- function term_ellipsis(s, maxwidth)
+--     if term_width(s) > maxwidth then
+--         repeat s = s:sub(1, #s - 1)
+--         until  #s == 0 or term_width(s) <= maxwidth - 3
+--         s = s .. (s:find("\27") and "\27[m..." or "...")
+--     end
+--     return s
+-- end
+
+-- if maxwidth <= 2 and s doesn't fit then the output is "..." (more than 2)
+-- function term_ellipsis(s, maxwidth)
+--     if term_width(s) > maxwidth then
+--         -- binary search for the longest prefix which fits maxwidth
+--         local low, high = 0, #s  -- low = 0 in case s is "\t..."
+--         while low < high-1 do
+--             local mid = (low + high) / 2
+--             if term_width(s:sub(1, mid)) <= maxwidth - 3 then
+--                 low = mid
+--             else
+--                 high = mid
+--             end
+--         end
+--         s = s:sub(1, low)
+--         s = s .. (s:find("\27") and "\27[m..." or "...")
+--     end
+--     return s
+-- end
 
 local function scroll_content(content_lines, offset, max_lines, unknown_height)
     local max_offset = unknown_height and content_lines or content_lines - max_lines
@@ -869,11 +918,22 @@ local function scroll_content(content_lines, offset, max_lines, unknown_height)
     return offset, max_lines
 end
 
+local time_total = 0
+local time_n = 0
 local function compose_page(header, content, offset, max_lines, unknown_height)
     local from, to = scroll_content(#content, offset, max_lines - #header, unknown_height)
     -- page keybinding page does it's own truncation
     if not unknown_height and curr_page ~= o.key_page_4 then
-        truncate_lines(content, from, to)
+        local start = mp.get_time()
+        for i = from, to do
+            content[i] = term_ellipsis(content[i], 80)
+        end
+        time_n = time_n + 1
+        local diff = mp.get_time() - start
+        if time_n > 15 then
+            time_total = time_total + diff
+            print(time_total / (time_n - 15) * 1000)
+        end
     end
     return table.concat(header) .. table.concat(content, "", from, to), from
 end
```
</details>
by commenting out the relevant parts, running `mpv --vid=no <path to video file> --force-window=no --terminal`, opening stats with `I` and then pressing arrow up to cause a bunch of refreshes. It's not `--no-config` because I wanted to have some long audio filters in there. To get the "starts at" measurement for luajit I've looked at the first output from a bunch of starts and picked a representative value (very scientific :D).
<details>
<summary>The stats output looked like this</summary>

```
File: Elephante - Come Back For You (feat. Matluck) 30.mkv
        Title: Elephante - Come Back For You (feat. Matluck)
        Size: 27.436 MiB        Format/Protocol: mkv
        Total Cache: 8.27 MiB (200.7 sec)

Audio: opus (Opus)
        Channels: 2     Format: floatp
        Sample Rate: 48000 Hz
        Bitrate: 151 kbps
        Filters:
                @afftdn: afftdn (disabled) [tn=1]
                @afwtdn: afwtdn (disabled)
                @anlmdn: anlmdn (disabled)
                @acompressor1up: acompressor (disabled) [mode=upward]
                @acompressor1down: acompressor (disabled) [mode=downward]
                @dynaudnorm: dynaudnorm [f=21 m=10 r=1.0 g=11]
                @noisegate: compand (disabled) [@3=.01 @4=0 @1=.2|.2 @2=-900/...
                @acompressor2down: acompressor (disabled) [mode=downward]
                @acompressor2up: acompressor (disabled) [mode=upward]
                @acompressorcdown: acompressor (disabled) [mode=downward thre...
                @acompressorcup: acompressor (disabled) [mode=upward threshol...
                @acontrast: acontrast (disabled)
```
</details>

Results
```
lua52
n2    takes 11ms
nlogn takes  1.84ms
n     takes  0.65ms

luajit
n2    starts at 5.2ms,  goes down to 4.46ms
nlogn starts at 1.36ms, goes down to 0.755ms
n     starts at 0.48ms, goes down to 0.0223ms
```

christoph-heinrich commented:
We decided to go with @avih's O(n) solution because it's more readable and more robust :tada:

avih commented:
Please stop adding more commits, or, alternatively, tell me when it's ready for review.

It's already at 7 commits and touching nearly 250 lines of delicate code.

It needs to be reviewed as well...

avih commented:
> made the persistent_overlay limit scrolling with compute_bounds.

Don't. this doesn't belong here, and I think it doesn't belong anywhere. At the very least remove it from this PR.

At the other commits, please ensure the commit messages describes well the goal of the change and any special notes about the the change.

kasper93 commented:
> Also, you didn't reply yet why do we actually need scrolling at the default page.

This is explained in the original bug report, no? https://github.com/mpv-player/mpv/issues/11544

avih commented:
> This is explained in the original bug report, no? #11544

No. It mentions something something, the default page overflows.

It doesn't explain why we need the default page scrollable, and doesn't have any argument of why the best place for that bunch of perf data is the default page.

It was discussed on IRC, and mentioned that the default page should remain summary if possible, without tons of additional info.

Detailed and long info data would probably be best served elsewhere.

kasper93 commented:
There is already dedicated page for perf data. There is zero need for `print_perfdata_passes`, but if there is an options, someone will use it and it should function correctly.

> Making it scrollable is one possible solution, not the only one, and not necessarily the best one.

Given that other pages are scrollable, this is very intuitive idea to make default page scrollable too if it overflows. It only affect users that enabled `print_perfdata_passes` anyway.

As I understand you argue that making default page scrollable is bad idea, why?

avih commented:
> if there is an options, someone will use it and it should function correctly.

Obviously true, and that was never disputed.

> this is very intuitive idea to make default page scrollable too if it overflows. That's probably why user suggested it

Obviously, and it's not a bad suggestion at all. But here we commit code, and when we do that we should evaluate whether it's indeed the best solution.

> but to accommodate corner cases like `print_perfdata_passes` it is a solution that does not harm anything, when not used.

Maybe, if that was a trivial change, but it's not.

The scroll works by accepting a list of lines, which is best suited for outputs which are long lists of whatever, like pages 0 or 4.

The default page is the opposite of that. It's a bunch of unrelated things, each outputs a bulk of data, and to make that reasonably scrollable, a lot of work is required.

So, if we can solve that by ensuring that the default page never needs scroll in the vast majority of use cases, then that might be preferrable.

Also, keep in mind that we don't _have_ to expose all the internal data at the stats page, and we simply can't expose everything. It's best used for data which many users would find valuable. For other things, user can write scripts which exports as much data as they want in whatever formats.

That's not an argument that we should remove the perfdata or filters copmpletely. It's just a piece of info which some might have not considered.

kasper93 commented:
> That's not an argument that we should remove the perfdata or filters copmpletely. It's just a piece of info which some might have not considered.

Sure, but I would in fact argue that removing `print_perfdata_passes` is actually best solution to avoid default page exploding. For filter list, I think it is quite hard to make it so long to become a problem. While VO passes are easy to make long. 

Note that we have dedicated page 2 for VO passes. So nothing is lost and I in fact have no idea why `print_perfdata_passes` option even exists. It really does not fit at all on page 1 (default)

```
diff --git a/player/lua/stats.lua b/player/lua/stats.lua
index 367a2af103..ab5db39eef 100644
--- a/player/lua/stats.lua
+++ b/player/lua/stats.lua
@@ -27,7 +27,6 @@ local o = {
     redraw_delay = 1,                -- acts as duration in the toggling case
     ass_formatting = true,
     persistent_overlay = false,      -- whether the stats can be overwritten by other output
-    print_perfdata_passes = false,   -- when true, print the full information about all passes
     filter_params_max_length = 100,  -- a filter list longer than this many characters will be shown one filter per line instead
     debug = false,
 
@@ -800,7 +799,7 @@ local function add_video(s)
     end
 
     append_display_sync(s)
-    append_perfdata(s, o.print_perfdata_passes)
+    append_perfdata(s, false)
 
     append_resolution(s, r, "Native Resolution:")
     if ro and (r["w"] ~= ro["dw"] or r["h"] ~= ro["dh"]) then
```

christoph-heinrich commented:
> Please stop adding more commits

You brought up using term_ellipsis for page 4, the scroll hint wasn't fixed earlier because I accidentally removed it in the rebase and thus didn't notice it having baked in ass tags, and the scroll limitation is now dropped again.

> tell me when it's ready for review

It's ready for review.

> > made the persistent_overlay limit scrolling with compute_bounds.
> 
> Don't. this doesn't belong here, and I think it doesn't belong anywhere. At the very least remove it from this PR.

I wasn't sure if that was wanted or not, but it turned out to be simple to do. Dropped.

> Also, please don't change default values quietly, like the 20 to 23 terminal llines. If you want, change default values in it's own commit, but please keep this one 20.

The comment said 20, but in reality it took up 22 lines of the terminal, leaving one empty line at the top in a 80x24 terminal.

> Also, you didn't reply yet why do we actually need scrolling at the default page.

I did say that it's overflowing for me because of filters and I also mentioned `print_perfdata_passes=yes`.

> We discussed it on IRC and we've concluded that you need either dozens of filters to make the default page overflow, or some other non-default things, and that it's highly unlikely to require scrolling for most users, AND that it's a very non-heterogeneous page which needs a lot of work to make it scrollable.

A lot of work in the sense that it's a lot of lines changed, but the change to make a output line into one entry is simple and merely repeated ton of times. And that trick with the temporary table for that line has also been done for page 0, so it's not like it introduces any new ideas.

> All those seem to point natively to move the filters and possibly other long output to a different page, and yet this commit still makes the default page scrollable.

I'm not going to drop a commit until we're sure it's not needed anymore, that's why the default page is still scrollable.
But I do agree that making it scrollable might not be the best way forward.

> We need to first agree that this is actually the best solution, and for now I don't think it is.

Then how about we leave the default page out of this PR and any changes to it will be something separate.

avih commented:
Thanks for the changes.

First commit looks good now.

Reviewing the rest...

avih commented:
2nd commit looks OK. Thanks.

avih commented:
Fourth commit (scroll hint) looks OK, but unrelated to the unification it has two issues:
1. The up/down arrows are not necessarily available at the font.
2. The scroll keys are configurable, and the hint ignores that.

So I think it would be useful to add a commit which makes the hint show the actual configured scroll key names, which would address both issues in one go - because the key names are likely to have the font glyphs available (even if it's some unicode key, the user is likely to have that glyph on their system).

As for the third commit (terminal truncation), few comments:
- It's only applied during scroll composition, so the default page is not truncated.
- It's hardcoded to truncate, so if the user has a bigger terminal they don't have a way to override it.
- I think it would be better to default to no-truncation, and allow truncation to a configurable value. This would apply to page 4 too, also in "stand alone" mode.

So maybe a useful set of commit would be:
1. Change the commit to apply to the default page too. Because the output table is not necessarily one line per element, it might be simpler to split the final text output on newlines, apply the truncation, and join it back together.
2. Change page 4 to use the unified function and ensure it works in stand alone too.
3. Add a truncation config option which defaults to no-truncation (e.g. maybe default 0 and values <= 0 disable truncation). Make sure stand alone page 4 adheres to it as well, e.g. maybe by applying the width value to that config option before generating the output. Don't forget to update the docs for the config option and page 4 CLI option.

Sounds reasonable? Other suggestions?

christoph-heinrich commented:
I feel like this will never get in, even though it's already a clear improvement over the status quo...

> Fourth commit (scroll hint) looks OK, but unrelated to the unification it has two issues:
> 
>     1. The up/down arrows are not necessarily available at the font.
> 
>     2. The scroll keys are configurable, and the hint ignores that.
> 
> 
> So I think it would be useful to add a commit which makes the hint show the actual configured scroll key names, which would address both issues in one go - because the key names are likely to have the font glyphs available (even if it's some unicode key, the user is likely to have that glyph on their system).

I agree that dynamically choosing the symbol(s) for the keybinding would be an improvement, but keep in mind I'm not the one that introduced that hint. Also that's mainly useful to new users, which will use have the default keybindings. A user that changes those keybindings certainly already knows about the scrolling.

> As for the third commit (terminal truncation), few comments:
> 
>     * It's only applied during scroll composition, so the default page is not truncated.
> 
>     * It's hardcoded to truncate, so if the user has a bigger terminal they don't have a way to override it.
> 
>     * I think it would be better to default to no-truncation, and allow truncation to a configurable value. This would apply to page 4 too, also in "stand alone" mode.
> 
> 
> So maybe a useful set of commits would be:
> 
>     1. Change the commit to apply to the default page too. Because the output table is not necessarily one line per element, it might be simpler to split the final text output on newlines, apply the truncation, and join it back together.
> 
>     2. Change page 4 to use the unified function and ensure it works in stand alone too.
> 
>     3. Add a truncation config option which defaults to no-truncation (e.g. maybe default 0 and values <= 0 disable truncation). Make sure stand alone page 4 adheres to it as well, e.g. maybe by applying the width value to that config option before generating the output. Don't forget to update the docs for the config option and page 4 CLI option.

You want bindlist to not specify the width anymore?

> Sounds reasonable? Other suggestions?

If it doesn't truncate by default, then users might not realize that this is even an option, where as if it does truncate by default and they don't like the width, they're more likely to look for a way to change it.

christoph-heinrich commented:
The new property is called `term-size` instead of `terminal-size` to make it consistent with the already existing options.
I didn't make it an option because I can't be bothered to anymore. In theory we could even expose `term-status-msg` and then count the newlines and adjust the output height accordingly, but that's an exercise for somone else... or maybe me at some point in the future, but not in this PR.

avih commented on player/lua/stats.lua:
```diff
@@ -867,24 +867,90 @@ local function eval_ass_formatting()
+
+-- split str into a table
+-- example: local t = split(s, "\n")
+-- init: where to start looking for the separator pattern (default 1)
```
Even though the `init` comment is mine and it's not wrong, it's also misleading, because the behavior is as if `str` was `str:sub(init)` (I added this init comment quickly right before I posted this function. my bad).

Also, because `init` is never used by callers, and currently it's unlikely to be used, I suggest to remove the `init` argument and always make it `1` inside the function.

`plain` can stay because splitting on a pattern can be useful, even if currently unused.

avih commented on player/lua/stats.lua:
```diff
@@ -867,24 +867,90 @@ local function eval_ass_formatting()
-    local to = min(#content, from + max_content_lines - 1)
-    return table.concat(header) .. table.concat(content, "", from, to), from
+-- max_lines: (optional) limit number of output lines to this,
+--            values <= 0 or nil mean unlimited and scrolling is ignored
```
If I'm reading it right, this comment is incorrect, because if `max_lines` is `nil` then it becomes 22 or 40.

Also, the word "scroll" is never mentioned elsewhere, so it's unclear what "scrolling is ignored" means. I think you mean "offset is considered 1" ? (why?)

Also, it looks to me like `compose_page` is always called with empty `header`, offset 1, and `nil` `max_lines`, which means:
- The output will be more than 22/40 elements if the header is not empty.
- The default page is limited to 22 terminal-lines / 40-array-elements eventhough it's not scrollable. That's a problem I think.
- Some of the argument might be redundant if they're unused and/or always used with the same value?

avih replied:
> Also, it looks to me like `compose_page` is always called with empty `header`, offset 1, and `nil` `max_lines`

Apologies, my bad.

That's how it's called at the new places which this commit adds, which are the non-scrollable default page, and two places at teh non-scrollable cache page. amd it's fine, because they're not scrollable.

The other places where it's called is pages 0, 2, and 4, where it is called with useful parameters.

However, I still think we can consider simplifying the interface to `finalize_page(header, content, apply_scroll)`. Thoughts?

christoph-heinrich replied:
We need max_width for bindlist, but max_lines can be replaced with a boolean.

avih replied:
Not if we have an override option which bindlist can set.

christoph-heinrich replied:
Yes, I should have tried actually implementing it before commenting, then I might have noticed.

avih commented on player/command.c:
```diff
@@ -2769,6 +2770,20 @@ static int mp_property_osd_ass(void *ctx, struct m_property *prop,
+static int mp_property_term_size(void *ctx, struct m_property *prop,
+                                  int action, void *arg)
+{
+    int w = 80, h = 24;
```
I prefer to fallback to 0,0, so that the caller can make different decisions when the terminal is 80x24 vs when the size is unknown.

Do check what other possibly-unknown numeric values fallback to, and if elsewhere uses `-1` for "unknown but returned anyway", then please use -1,-1 here too.

llyyr replied:
0 would signify an error, we generally return `M_PROPERTY_UNKNOWN` when some properties are unknown

christoph-heinrich replied:
Afaik `M_PROPERTY_UNAVAILABLE` is for when the property is currently not available, while M_PROPERTY_UNKNOWN is for properties that don't exist.

I do agree that we should return an error instead of a dummy value.

avih replied:
Technically we can't know whether `get)terminal_size` failed or not, and we have a precedence that `osd-dimentions` has 0 if unknown. So I prefer 0,0.

christoph-heinrich replied:
We can set w and h to -1 and then read the dimensions. I'm pretty sure a successful call would never return that.

avih commented on player/lua/stats.lua:
```diff
@@ -930,13 +930,13 @@ end
 -- max_width      : (optional) limit line width to that many characters
 -- unlimited_lines: (optional) don't limit output lines and ignore scrolling
 local function compose_page(header, content, offset, max_width, unlimited_lines)
+    local terminal_size = mp.get_property_native("term-size")
```
The property might not work on all platforms. This needs an option to allow override the automatic truncation.

Maybe even if it does work the user wants to never truncate, etc.

christoph-heinrich commented:
The property can be unknown now instead of returning dummy values.

The new option has the syntax WxH, where each dimension can be A for automatic, 0 for infinite or some positive integer. Defaulting to AxA.

christoph-heinrich commented:
The options are `term_width_limit` and `term_height_limit` now, with -1 (technically <0) being automatic and 0 is infinite.
Internally automatic is represented as nil because that makes fall back easier.

christoph-heinrich commented:
So far it wasn't possible to observe `term-size` because nothing signaled any terminal size changes.
I've implemented that in a [branch](https://github.com/christoph-heinrich/mpv/commits/term_size_updates) on top of this PR in case anyone cares. Maybe that should become part of cc4b589b2c031c2ed291f7855ff9bf3c2c6f6a67 and 7364602db58368638f395816d5152e68c16e2a0a?
It doesn't work on windows because afaict there is no way of doing it there besides constantly polling the size.

kasper93 commented on DOCS/man/stats.rst:
```diff
@@ -85,6 +85,18 @@ Configurable Options
     scripts decide to print text at the same time.
 
+``term_width_limit``
+    Default: -1
```
I don't think it is correct? Looking at
```
    if (w == -1 || h == -1)
        return M_PROPERTY_UNAVAILABLE;
```

christoph-heinrich replied:
This is an option for stats, not the propery.

kasper93 commented on player/lua/stats.lua:
```diff
@@ -82,6 +84,15 @@ local o = {
 
+o.term_width_limit = tonumber(o.term_width_limit) or -1
+o.term_height_limit = tonumber(o.term_height_limit) or -1
+if o.term_width_limit < 0 then
+    o.term_width_limit = nil
+end
```
nit, could also be a one liner

christoph-heinrich replied:
But that's not in line with the code style of the rest of the file.
Although there are two one liner `if`s, so there is precedent...

kasper93 reviewed (commented):
Generally looks ok. The terminal size thing is little tricky, but I guess it is ok.

kasper93 commented on DOCS/interface-changes.rst:
```diff
@@ -155,6 +155,7 @@ Interface changes
     - remove `border`, `fullscreen`, `ontop`, `osd-level` and `pause`
       from default `--watch-later-options`
     - add `video-*` and `secondary-sub-visibility` to default `--watch-later-options`
+    - add `term-size` property
```
please rebase, it is in 0.37 section.

christoph-heinrich replied:
Thanks for the heads up

kasper93 commented:
No need to keep changes like that open forever, merged, thanks.
